Light editor multi edit

ericmehl opened the pull request:
This adds the ability to select and edit multiple cells in one interaction. To accomplish this, it also replaces `allowMultipleSelection` with `selectionMode` in the `PathListingWidget`, with options for `Cell` and `Cells` in addition to the backwards compatible `Row` and `Rows`.

This is a re-PR of https://github.com/GafferHQ/gaffer/pull/4726 which was closed when I re-forked my personal repository onto HQ.

From that PR, I made the change to the `ScopedGILRelease` in `setSelection` which fixed the problem with using `extend_container`. And hopefully the crash that popped up in the CI tests.

I also looked again at what I thought was a problem with selection expansion. I think that the test I had been using wasn't quite correct. I had been setting the selection to include `/4/9` and expecting `/4` to be expanded which I believe is not the intended behavior. The test is changed to reflect the expected behavior and is now passing.

Ready for reviewing of the code and all the key and click combinations we can think of to try and find a hole in the implementation.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -792,33 +862,33 @@ def drawRow( self, painter, option, index ) :
-			color.setAlpha( alpha )
-			painter.fillRect( rect, color )
+			color.setAlpha( 50 )
+			painter.fillRect( activeRect, color )
```
We should be painting `rect` here, to paint the full row. If the last element in the row is selected, then `activeRect` is just the rect for the last cell. This causes the issue below, where we lose the descendant highlighting for the top row...

![image](https://user-images.githubusercontent.com/1133871/176470455-f7bf0e43-6251-4bd5-80e4-c0be5085b1e1.png)

ericmehl replied:
Fixed in 8800d2172415d8bdcad82244c1be89edc2341afb

johnhaddon commented on src/GafferUIModule/PathListingWidgetBinding.cpp:
```diff
@@ -1761,20 +1794,32 @@ IECore::PathMatcher getExpansion( uint64_t treeViewAddress )
+
+	list result;
+
+	PathModel::Selection selection = model->getSelection();
```
Is there a reason we're taking a copy here? Might it be cheaper and clearer just to do `for( const auto &p : model->getSelection() )`?

ericmehl replied:
Nope, fixed in 60336bfa71fa897f245bdb9517d56294bfc4d5cc

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -319,6 +320,57 @@ def __buttonDoubleClick( self, pathListing, event ) :
+			nonEditable = [ i for i in inspections if not i.editable() ]
+
+			if len( nonEditable ) == 0 :
+				edits = [ i.acquireEdit() for i in inspections ]
```
The plugs acquired here can be of different types, in which case we get an exception like this :

```
File "/home/john/dev/build/gafferPython3/python/GafferSceneUI/TweakPlugValueWidget.py", line 82, in __init__
    self.__row.append( GafferUI.PlugValueWidget.create( self.__childPlugs( plugs, "value" ) ), expand = True )
  File "/home/john/dev/build/gafferPython3/python/GafferUI/PlugValueWidget.py", line 252, in create
    raise Exception( "Multiple widget creators" )
Exception: Multiple widget creators
```

In the Spreadsheet, we have a PopupWindow saying "Unable to edit plugs with mixed types". I think we need to do something similar here.

ericmehl replied:
It's perhaps not the best solution, but I added a dialog in 029cbf28f4dd2bbdbd080496029ffaa85503f21e warning of mismatched plugs.

I included a todo that I think might be an improvement that can be made on `0.62` moving the handling of `TweakPlug` with different value types into `PlugPopup` that might be a bit cleaner.

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -319,6 +320,57 @@ def __buttonDoubleClick( self, pathListing, event ) :
+				self.__popup.popup()
+			else :
+
+				# See todo in `PlugPopup._PopupWindow`
+				PopupWindow = GafferUI.PlugPopup.__bases__[0]
```
Note to self : this is all sorted on `main` now, so you'll need to sort it during the merge.

johnhaddon commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -319,6 +320,57 @@ def __buttonDoubleClick( self, pathListing, event ) :
 
 		return True
 
+	def __keyPress( self, pathListing, event ) :
```
Most of this function feels like a multi-selection-aware superset of `__buttonDoubleClick()`. Since the double-clicked item should always be selected, can we just delegate both click and keypress to a single `__editSelectedCells()` method?

ericmehl replied:
Yes, much better. Fixed in 05cf74f11693462845f662bb682f6a294d96fd86

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -587,7 +584,6 @@ def __keyPress( self, widget, event ) :
 	# for compatibility with `GafferUI.Widget.dragBeginSignal()`.
 	def __buttonPress( self, widget, event ) :
 
-		self.__updateSelectionInButtonRelease = False
```
I haven't reviewed this commit thoroughly, as I think it is problematic. I think it is the reason that it's no longer possible to drag a multiple selection from the hierarchy view. The click to start the drag is immediately deselecting all the other paths...

ericmehl commented on src/GafferUIModule/PathListingWidgetBinding.cpp:
```diff
@@ -409,6 +414,15 @@ class PathModel : public QAbstractItemModel
+				rowSelection = IECore::PathMatcher( m_selection[0] );
+			}
+
+			setSelection( Selection( columns.size(), IECore::PathMatcher() ) );
+			m_selection[0] = rowSelection;
```
Yes, that's a logic error on my part. The motivation is from the Light Editor, where switching columns could be common and I think it would be desirable to keep the selected lights. I still think that's worth keeping, especially for the Light Editor since the first column also controls scene selection.

I could be persuaded otherwise, perhaps there are other factors I'm not considering.

johnhaddon replied:
I like the LightEditor behaviour - I think perhaps this just isn't the right place to implement it.

For Row/Rows mode, it seems reasonable that PathListingWidget should automatically maintain selection in `setColumns()`, because nothing about changing columns changes the interpretation of the selection. So we can just copy `m_selection[0]` to all columns.

But for Cell/Cells mode I don't think PathListingWidget can reason about the new columns at all, so clearing selection makes sense. Unless...maybe we should detect when an old column is reused in `setColumns()` and maintain the selection for any of those? I think this would actually give us the behaviour we want in the LightEditor and doesn't rely on PathListingWidget having any external knowledge of anything.

ericmehl replied:
I pushed a new commit fa80d567e6c01842f2d280f6e71384a4055ef096 which I think accomplishes this correctly. `LightEditorInspectorColumn` didn't seem to have a `cellData` binding, so I added one in that commit.

johnhaddon replied:
Yep, I don't think `PathModel` has any choice but to compare by pointer identity, because Column doesn't have an equality operator. For clients, the rule will be "if you want to maintain selection, reuse existing columns, _don't_ make new columns that happen to be the same". I think that's pretty reasonable...

ericmehl commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -318,24 +318,59 @@ def getSortable( self ) :
+		if (
+			self.__selectionMode == self.SelectionMode.Row or
+			self.__selectionMode == self.SelectionMode.Cell
+		) and sum( [ p.size() for p in  set( [ p for p in paths ] ) ] ) > 1 :
```
Fixed in b9e5707e4dc8f98e77882aebe5b2e1093285d36e

ericmehl commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -318,24 +318,59 @@ def getSortable( self ) :
-		assert( isinstance( paths, IECore.PathMatcher ) )
-		if self.__selectionMode == self.SelectionMode.Row and paths.size() > 1 :
-			raise ValueError( "More than one path selected" )
+		assert( isinstance( paths, list ) or isinstance( paths, IECore.PathMatcher ) )
```
Fixed in b9e5707e4dc8f98e77882aebe5b2e1093285d36e

ericmehl commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -318,24 +318,59 @@ def getSortable( self ) :
+			self.__selectionMode == self.SelectionMode.Rows and
+			len( set( [ p for p in paths ] ) ) > 1
+		) :
+			raise ValueError( "Selection must be the same for all columns" )
```
Fixed in b9e5707e4dc8f98e77882aebe5b2e1093285d36e

ericmehl commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -318,24 +318,59 @@ def getSortable( self ) :
+	#
+	# Within `PathListingWidget` we standardise on using a
+	# list of `PathMatcher`. Returning a single `PathMatcher`
+	# is provided for compatibility.
```
This has jumped around a bit in the fixup history, first in b9e5707e4dc8f98e77882aebe5b2e1093285d36e and then in 0b261c2f161a3070da5508b7a370ad14a9946cf4 and 1c2113945263d7f1b589a86905b1fe88d91d661b

ericmehl commented on src/GafferUIModule/PathListingWidgetBinding.cpp:
```diff
@@ -387,6 +387,11 @@ class PathModel : public QAbstractItemModel
+		// Selection is stored as a single `IECore::PathMatcher` per column, allowing
+		// per-cell selections. An empty selection is represented by a vector of empty
+		// `PathMatcher()` equal in size to the number of columns.
+		typedef std::vector<IECore::PathMatcher> Selection;
```
Fixed in c8dfe556759404543ea2b5abbe9179308726a39e

ericmehl commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -792,33 +862,33 @@ def drawRow( self, painter, option, index ) :
-		# partially-transparent backgrounds to be drawn too heavy. Refill with the
-		# default background colour before drawing our highlight and deferring the
-		# rest to Qt.
-		painter.fillRect( rect, QtGui.QColor( *(_styleColors["backgroundRaised"]) ) )
```
Fixed in 8800d2172415d8bdcad82244c1be89edc2341afb

ericmehl commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -792,33 +862,33 @@ def drawRow( self, painter, option, index ) :
+				descendantMatch = True
 
-	def __drawSelectionHighlight( self, painter, rect, index ) :
+			activeRect = rect
```
Fixed in 8800d2172415d8bdcad82244c1be89edc2341afb

ericmehl commented on Changes.md:
```diff
@@ -21,7 +21,10 @@ Improvements
 - CameraTweaks : Added `Create` mode. This should be used in preference to `Replace` mode when the parameter may not exist.
 - PathFilter : Added the ability to select scene locations from currently selected filter entries.
-- PathListingWidget : Added an option to `selectionMode` allowing single and multiple cell selections.
+- PathListingWidget :
+  - Added an option to `selectionMode` allowing single and multiple cell selections.
+  - Added the ability to select cells in different columns with the left <kbd>&larr;</kbd> and right <kbd>&rarr;</kbd> keys.
+  - Added the ability to drag from a cell and drop that cell's value on an appropriate target.
```
Not exactly fixed, but called out so I won't forget it when rebasing at the end, in f6727d734a9a846297e7fd7fe18f57cd85b75907

ericmehl replied:
This is taken care of again, individually in the latest commits.

ericmehl commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -732,9 +731,15 @@ def __dragBegin( self, widget, event ) :
 		path = self.pathAt( imath.V2f( event.line.p0.x, event.line.p0.y ) )
 		selection = self.getSelection()
-		if selection.match( str( path ) ) & IECore.PathMatcher.Result.ExactMatch :
+		selection = selection if isinstance( selection, list ) else [selection] * len( self.getColumns() )
```
I agree, internal methods for those are nice to have. I added them in 0b261c2f161a3070da5508b7a370ad14a9946cf4

ericmehl commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -732,9 +731,15 @@ def __dragBegin( self, widget, event ) :
+		if index is not None :
 			GafferUI.Pointer.setCurrent( self.__dragPointer )
-			return IECore.StringVectorData( selection.paths() )
+			return self._qtWidget().model().data( index, QtCore.Qt.DisplayRole )
```
I pushed a new commit fa80d567e6c01842f2d280f6e71384a4055ef096 which I think accomplishes this correctly. `LightEditorInspectorColumn` didn't seem to have a `cellData` binding, so I added one in that commit.

And the icon is fixed in 029cbf28f4dd2bbdbd080496029ffaa85503f21e

ericmehl commented:
I squashed down the commits after addressing the remaining comments. They're somewhat monolithic commits now, unfortunately - the changes and refactors were crossing a number of previous commits to make them reliant on each other. Hopefully not too hard to review, or I can make edits if it would be better.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -800,24 +914,50 @@ def drawBranches( self, painter, rect, index ) :
 		QtWidgets.QTreeView.drawBranches( self, painter, rect, index )
 
-	def __drawSelectionHighlight( self, painter, rect, index ) :
+	def __drawSelectionHighlight( self, painter, rect, index, fullRow = False ) :
```
I haven't followed the new `fullRow` logic here fully, but something is off :

![image](https://user-images.githubusercontent.com/1133871/177303392-28078da0-3dcb-4e7e-a15c-199b21b295e4.png)

Both `/group` and `/group/spotLight` are selected, but the descendant selection is clobbering the full selection for all of the row except the branches section.

ericmehl replied:
My logic behind `fullRow` is that when drawing the branch, we shouldn't iterate over the columns because the `rect` given to us covers only the left-most width of the branch area.

If we do iterate over columns when drawing the branch area and the last column is selected, the `width` calculated will be negative (because `rect.right()` is the width of the branch area) and the whole row (minus the branch area) is drawn. 

In theory we could also use it for drawing in `Row` and `Rows` modes, but I believe that would require a new method like `selectionMode()` or similar for `_TreeView` to be able to know what mode it's in. I avoided that to keep the overall changes down.

I'm open to clearer naming or better logic of course.

ericmehl replied:
Fixing the highlight behavior made it a bit easier to organize the branch vs. row data logic, so the `fullRow` bit is no longer needed. Fixed in b318f952c56fffaa60bcd2eb0c92ae1b6e35cc29

johnhaddon commented on Changes.md:
```diff
@@ -55,6 +70,9 @@ API
 - EditScopeAlgo : Added support for editing attributes.
 - TweakPlug : Added `Create` mode.
 - VectorDataWidget : Added `dataMenuSignal` for constructing the context menu. Classes can add their own slots to this signal to modify the context menu. Derived classes should use it in favor of `_contextMenuDefinition()`.
+- PathListingWidget :
+  - Deprecated `allowMultipleSelection` constructor argument. Use `selectionMode` argument instead.
+  - `setSelection` and `getSelection` can now accept either a single `PathMatcher` that represents the selection for all columns, or a list of `PathMatcher` objects, one for each column. If `selectionMode` is `Row` or `Rows`, `getSelection` returns a single `PatchMatcher` for backwards compatibility.
```
This is under the wrong version, and the documentation for `setSelection()` and `getSelection()` doesn't quite reflect our tweaked behaviour for those functions any more.

johnhaddon commented on Changes.md:
```diff
@@ -38,6 +48,11 @@ Improvements
 - Light Editor : Added support for editing visualisation attributes of lights.
 - CameraTweaks : Added `Create` mode. This should be used in preference to `Replace` mode when the parameter may not exist.
 - PathFilter : Added the ability to select scene locations from currently selected filter entries.
+- PathListingWidget :
+  - Added an option to `selectionMode` allowing single and multiple cell selections.
+  - Added the ability to select cells in different columns with the left <kbd>&larr;</kbd> and right <kbd>&rarr;</kbd> keys.
+  - Added the ability to drag from a cell and drop that cell's value on an appropriate target.
+- Light Editor : Added the ability to edit multiple values at one time. With multiple cells selected, pressing <kbd>Return</kbd> or <kbd>Enter</kbd> will open a popup to edit the values for all cells.
```
This is under the wrong version, and is a duplicate.

johnhaddon commented on Changes.md:
```diff
@@ -1,6 +1,16 @@
+Improvements
+------------
+
+- PathListingWidget :
+  - Added an option to `selectionMode` allowing single and multiple cell selections.
```
This belongs in the API section. And maybe the interaction descriptions belong in the LightEditor section, since that's the only place users will see them right now?

johnhaddon commented on Changes.md:
```diff
@@ -1,6 +1,16 @@
+  - Added an option to `selectionMode` allowing single and multiple cell selections.
+  - Added the ability to select cells in different columns with the left <kbd>&larr;</kbd> and right <kbd>&rarr;</kbd> keys.
+  - Added the ability to drag from a cell and drop that cell's value on an appropriate target.
+  - Improved the range selection behavior when using <kbd>Shift</kbd> + arrow keys. It now selects cells in a similar way to <kbd>Shift</kbd> + Click.
```
I don't think this has changed from a user perspective - the old rows-only behaviour is the same as before, and the cell modes are brand new.

johnhaddon reviewed (changes requested):
Thanks Eric! I've had a good play with this and the new consolidated-keypress-and-buttonpress selection code all seems to be working nicely. I've made a couple of new comments inline that we'll need to sort out before we can merge, but I think we can get this in before we release 1.0.0.0?

ericmehl commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -269,41 +272,78 @@ def __selectionChanged( self, pathListing ) :
+			# Once `TweakPlug` is in the `Gaffer` module, could this be handled
+			# in `PlugPopup`? It would make the UI consistent with
+			# `Spreadsheet` and less intrusive to the user.
+			if sole( i.source().__class__ for i in inspections ) :
```
I think I'm following the logic here in d637a3a0813c98212d09b8c413eab3b73bd46875 and dab0f048319624871d8ce73c63fafded9aedbe4d, with the exception of the circular reference. Is that not handled by https://github.com/GafferHQ/gaffer/blob/8e683113ad4cc1c92ca9648a4ce8d0ff05aea9f6/python/GafferUI/PlugValueWidget.py#L794-L795 or is there another circular reference I'm introducing?

ericmehl replied:
An update to the above comment - after rebasing and some tweaks, this is fixed in ec4b301366288e439802cc1c3eb702f1a01888ba, 98be6b1097db082f96010c31f464e4a40b76ca6e, 3bf87a773ad06e317efcae83a0936bff0e7ddf6f and 94b3174d47608b86f3d1b760c78ca71114f92ff6. Keeping them separate for review purposes, but they can probably be squashed before merging.

Adding `MultipleWidgetCreatorsError` was a result of `create()` checking for mismatching widget creators before the check for mismatching types. `MultiplePlugTypesError` may not be strictly needed for this? I added it to be (very nearly) consistent with the exceptions.

ericmehl commented on src/GafferSceneUIModule/LightEditorBinding.cpp:
```diff
@@ -242,6 +242,7 @@ void GafferSceneUIModule::bindLightEditor()
 			)
 		) )
 		.def( "inspector", &InspectorColumn::inspector, return_value_policy<IECorePython::CastToIntrusivePtr>() )
+		.def( "cellData", &InspectorColumn::cellData, ( arg_( "path" ), arg_( "canceller" ) = object() ) )
```
I think this is fixed in 81f350e0144954d201f9a3dede93e71e06cff1ad. When we discussed offline you mentioned one case concerning derived classes in Python that wouldn't get their `cellData` called. I'm still a bit fuzzy on that code path and how to handle it, but it also sounded like it may be unnecessary?

ericmehl commented:
Ok, I think everything is handled now, except potentially the `cellData` binding. The `Changes.md` log is fixed and tidied up considerably.

Ready for a fresh review...

ericmehl commented:
I squashed down to I think a reasonable history and should be ready for merging!
